### PR TITLE
[PyTorch][ET] Collect Process Groups Mapping Info

### PR DESCRIPTION
**Pull request by louisfeng**
Summary: Add the logics and interface to log ProcessGroup comms configuration (unique ID, type, and ranks info).

Test Plan:
Testing in HPC:
```
TORCH_LOGS=all ../buck-out/v2/gen/fbcode/c8344b52091f4f7f/hpc/models/ads/__ads_10x_launcher__/ads_10x_launcher.par  +launcher=local launcher.num_trainers=4 +data_loader=random data_loader.num_batches=2000
```
Example output in ET:
```
    {
      "name": "## process_group:init ##", "id": 3, "rf_id": 1, "parent": 2, "fw_parent": 0, "seq_id": -1, "scope": 7, "tid": 1, "fw_tid": 0, "op_schema": "",
      "inputs": ["[{'pg_id': 140538064364672, 'backend_id': 140538060772480, 'backend_config': 'cuda:nccl', 'ranks': {0: 0, 1: 1, 2: 2, 3: 3}}, {'pg_id': 140538064363904, 'backend_id': 140538042628864, 'backend_config': 'cuda:nccl', 'ranks': {0: 0, 1: 1, 2: 2, 3: 3}}]"], "input_shapes": [[]], "input_types": ["String"],
      "outputs": [], "output_shapes": [], "output_types": []
    },
```

Differential Revision: D46321690

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -1375,6 +1375,11 @@ that adds a prefix to each key inserted to the store.
           .def("rank", &::c10d::ProcessGroup::getRank)
           .def("size", &::c10d::ProcessGroup::getSize)
           .def("name", &::c10d::ProcessGroup::getBackendName)
+          .def("id", &::c10d::ProcessGroup::getID)
+          .def(
+              "backend_id",
+              &::c10d::ProcessGroup::getBackendID,
+              py::arg("backend_type"))
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/init.cpp:1382]: I wonder if these APIs must be exposed to end users?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -513,6 +520,22 @@ def pg_coalesce_state(self) -> Dict[ProcessGroup, List[Union[_CollOp, P2POp]]]:
     def pg_default_device(self) -> Dict[ProcessGroup, torch.device]:
         return self._pg_default_device
 
+    @property
+    def pg_config_info(self) -> List[Dict[str, str]]:
```
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:524]: Where is this function used?

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -513,6 +520,22 @@ def pg_coalesce_state(self) -> Dict[ProcessGroup, List[Union[_CollOp, P2POp]]]:
...
+        config_info = []
+        for pg, backend in self.pg_map.items():
+            backend_type = Backend.backend_type_map[backend[0]]
```
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:527]: why backend[0]? Some clarification would be helpful if you think [0] is the right behavior.

File: torch/_C/_autograd.pyi [hunk 4]
```diff
@@ -91,7 +91,7 @@ def _profiler_enabled() -> bool: ...
 def _add_metadata_json(key: str, value: str) -> None: ...
 def _kineto_step() -> None: ...
 def kineto_available() -> bool: ...
-def _record_function_with_args_enter(name: str, args: List[Any]) -> torch.Tensor: ...
+def _record_function_with_args_enter(name: str, *args) -> torch.Tensor: ...
```
Comment (kwen2501) [inline torch/_C/_autograd.pyi:94]: How is this change relate to the rest of the PR? (Is it necessary?)

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (louisfeng) [reply torch/distributed/distributed_c10d.py:524]: This is currently used in one of the trace collection step to inject this information into the trace so the traces can be used for network replay and triaging.

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (louisfeng) [reply torch/distributed/distributed_c10d.py:527]: I'll add some comments. `backend` is a tuple with the first element being the backend type (nccl, etc.).

File: torch/_C/_autograd.pyi (hunk 4 above)
Comment (louisfeng) [reply torch/_C/_autograd.pyi:94]: This is required for the internal diff (which has other code depends on this external diff). This fixes an edge case where `args` is a single value, which would work correctly with `*args` but not `args: List[Any]`. A workaround is to put the single value in a list, but that would change its type. `List[Any]` is a mismatch of `py::args args` semantics (see https://pybind11.readthedocs.io/en/stable/advanced/functions.html#accepting-args-and-kwargs)

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (kwen2501) [reply torch/distributed/distributed_c10d.py:524]: So this intends to be a user-facing API. Would you mind adding a short description under the def line? Thank you!

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (louisfeng) [reply torch/distributed/distributed_c10d.py:532]: Added comments.

File: torch/distributed/distributed_c10d.py (hunk 2 above)
Comment (louisfeng) [reply torch/distributed/distributed_c10d.py]: I have added comments to clarify the usage of this function.

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (louisfeng) [reply torch/csrc/distributed/c10d/init.cpp]: Added prefix.

**Review by kwen2501 — APPROVED** [review]
Thanks for the changes. Looks good to me.

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -3895,18 +3923,17 @@ def _new_group_with_tag(
 
     group_name = _process_group_name(ranks, use_hashed_name=use_local_synchronization)
 
-    with record_function(f"## process_group:init with ranks: {ranks}"):
```
Comment (fduwjj) [inline torch/distributed/distributed_c10d.py:3898]: Why do we don't need this for profiler?

**Pull request closed without merging**